# ALL NBA TEAMS PROJECT (2)

In this notebook I do the following:
 - Read cleaned data back from SQL
 - Test Hypothesis
 - Transform data, push it to SQL and save it as csv

In [1]:
# load datasets
import pandas as pd
import numpy as np

import pymysql                        # for getting data from a SQL database
from sqlalchemy import create_engine  # for establishing the connection and authentication

from getpass import getpass  # To get the password without showing the input

## Read data back from SQL

In [2]:
# I created a new schema in SQL and I'll save the cleaned dataframes there
password = getpass()

········


In [3]:
connection_string = 'mysql+pymysql://root:'+password+'@localhost/nba'
engine = create_engine(connection_string)

In [4]:
# read the data from SQL back to this notebook
player_stats = pd.read_sql_table('player_stats',engine)
display(player_stats.head())

quintets = pd.read_sql_table('all_nba_teams',engine)
display(quintets.head())

index                 player   season     team  pos  gs  as  age   g    mp  \
0      0            Bam Adebayo  2023-24      MIA  C-F  43   1   26  43  34.5   
1      1          Grayson Allen  2023-24      PHO    G  47   0   28  47  32.9   
2      2          Jarrett Allen  2023-24      CLE  C-F  46   0   25  46  30.5   
3      3  Giannis Antetokounmpo  2023-24      MIL  F-G  51   1   29  51  35.0   
4      4             OG Anunoby  2023-24  NYK,TOR    F  41   0   26  41  34.1   

   ...  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  
0  ...   1.4  19.5   7.1  27.9  17.3  19.4   1.5   2.8  11.9  26.0  
1  ...   1.1  13.5   2.4  11.1   7.0  12.5   1.3   1.7  12.7  13.9  
2  ...   2.0  22.1  13.2  25.0  19.3  13.4   1.3   3.4  13.4  18.9  
3  ...   4.6  29.2   8.2  26.4  17.6  30.5   1.9   2.5  12.9  33.1  
4  ...   0.7  13.5   3.1  10.6   6.8   9.4   1.9   1.8  10.7  17.6  

[5 rows x 52 columns]

index   season       first_team      second_team      third_team
0      0  1989-90      Karl Malone       Larry Bird    James Worthy
1      1  1989-90  Charles Barkley     Tom Chambers    Chris Mullin
2      2  1989-90    Patrick Ewing  Hakeem Olajuwon  David Robinson
3      3  1989-90    Magic Johnson    John Stockton   Clyde Drexler
4      4  1989-90   Michael Jordan    Kevin Johnson      Joe Dumars

In [5]:
# drop 'index' columns in both dataframes
player_stats = player_stats.drop(['index'], axis=1)
quintets = quintets.drop(['index'], axis=1)

In [6]:
display(player_stats)
display(quintets)

player   season     team  pos  gs  as  age   g    mp  \
0               Bam Adebayo  2023-24      MIA  C-F  43   1   26  43  34.5   
1             Grayson Allen  2023-24      PHO    G  47   0   28  47  32.9   
2             Jarrett Allen  2023-24      CLE  C-F  46   0   25  46  30.5   
3     Giannis Antetokounmpo  2023-24      MIL  F-G  51   1   29  51  35.0   
4                OG Anunoby  2023-24  NYK,TOR    F  41   0   26  41  34.1   
...                     ...      ...      ...  ...  ..  ..  ...  ..   ...   
5159      Dominique Wilkins  1989-90      ATL  F-G  79   1   30  80  36.1   
5160         Gerald Wilkins  1989-90      NYK  G-F  80   0   26  82  31.8   
5161          Buck Williams  1989-90      POR  F-C  82   0   29  82  34.2   
5162           Kevin Willis  1989-90      ATL  F-C  51   0   27  81  28.1   
5163           James Worthy  1989-90      LAL    F  80   1   28  80  37.0   

        fg  ...  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  
0      7.6  ...   1.4  19.5   7.1  27.9  17.3  19.4   1.5   2.8  11.9  26.0  
1      4.3  ...   1.1  13.5   2.4  11.1   7.0  12.5   1.3   1.7  12.7  13.9  
2      6.4  ...   2.0  22.1  13.2  25.0  19.3  13.4   1.3   3.4  13.4  18.9  
3     11.5  ...   4.6  29.2   8.2  26.4  17.6  30.5   1.9   2.5  12.9  33.1  
4      5.8  ...   0.7  13.5   3.1  10.6   6.8   9.4   1.9   1.8  10.7  17.6  
...    ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
5159  10.1  ...   5.4  24.6   8.4  12.4  10.4  11.8   2.2   1.0   8.3  30.4  
5160   5.8  ...   1.1  14.3   5.8   9.8   7.9  18.5   1.8   0.5  14.5  21.4  
5161   5.0  ...   1.9  14.3   9.6  21.6  15.5   5.5   1.2   0.8  15.3  15.5  
5162   5.2  ...   0.1  15.7  12.5  20.4  16.4   3.7   1.4   1.3  13.6  19.5  
5163   8.9  ...   4.3  19.8   6.4  11.9   9.3  15.6   1.7   1.0  10.0  23.0  

[5164 rows x 51 columns]

season               first_team       second_team        third_team
0    1989-90              Karl Malone        Larry Bird      James Worthy
1    1989-90          Charles Barkley      Tom Chambers      Chris Mullin
2    1989-90            Patrick Ewing   Hakeem Olajuwon    David Robinson
3    1989-90            Magic Johnson     John Stockton     Clyde Drexler
4    1989-90           Michael Jordan     Kevin Johnson        Joe Dumars
..       ...                      ...               ...               ...
165  2022-23    Giannis Antetokounmpo      Jimmy Butler      LeBron James
166  2022-23             Jayson Tatum      Jaylen Brown     Julius Randle
167  2022-23              Joel Embiid      Nikola Jokić  Domantas Sabonis
168  2022-23              Luka Dončić     Stephen Curry      De'Aaron Fox
169  2022-23  Shai Gilgeous-Alexander  Donovan Mitchell    Damian Lillard

[170 rows x 4 columns]

## Test Hypothesis

I select a value for alpha of 0.05 (p-value threshold, significance level)
One-sided test:
- H0: mean fg% (field goal %) value >= 45
- H1: mean fg% (field goal %) value < 45

In [7]:
player_stats_fg_percent = player_stats['fg%']

In [8]:
from scipy.stats import ttest_1samp
stat, pval = ttest_1samp(player_stats_fg_percent, 45)

print('stat is  ', stat) 
print('pvalue for the one-tailed test is ', pval/2)

stat is   25.692328698757354
pvalue for the one-tailed test is  2.1306768048709292e-137


- pvalue is very close to 0, so pvalue < 0.05. I have to check stat value. 
- stat value is positive and my H0 checks: mean fg% value >= 45. Since the sign is in the same way **I don't reject H0**.

stat value tells us that the true value is 25.69 times the mean square root error above 45.

In [9]:
player_stats['fg%'].mean()

46.878214562354756

## Transforming data

In [10]:
# split 'team' column into 3 columns and change column names
split_team_col = player_stats['team'].str.split(',', expand=True)
split_team_col= split_team_col.rename(columns={0:'playedfor_1',
                            1:'playedfor_2', 
                            2:'playedfor_3'})
split_team_col.head()

playedfor_1 playedfor_2 playedfor_3
0         MIA        None        None
1         PHO        None        None
2         CLE        None        None
3         MIL        None        None
4         NYK         TOR        None

I want to resume these 3 columns into 1 that just hold the information about how many teams has the player played for
in that season


In [11]:
# define a function to replace a value with 1 and a NaN value with 0
def fillplayedfor (x):
    if pd.isna(x):
        return 0
    else:
        return 1

In [12]:
# apply the function in all columns
for col in split_team_col.columns:
    split_team_col[col] = split_team_col[col].apply(fillplayedfor)

In [13]:
split_team_col['played_for'] = split_team_col['playedfor_1'] + split_team_col['playedfor_2'] + split_team_col['playedfor_3']
display(split_team_col.head())

playedfor_1  playedfor_2  playedfor_3  played_for
0            1            0            0           1
1            1            0            0           1
2            1            0            0           1
3            1            0            0           1
4            1            1            0           2

In [14]:
split_team_col = split_team_col.drop(['playedfor_1', 'playedfor_2', 'playedfor_3'], axis=1)
split_team_col.value_counts()

played_for
1             4806
2              352
3                6
Name: count, dtype: int64

In [15]:
# split 'pos' column into 3 columns and change column names
split_pos_col = player_stats['pos'].str.split('-', expand=True)
split_pos_col = split_pos_col.rename(columns={0:'pos_1',
                            1:'pos_2'})
split_pos_col.head()

pos_1 pos_2
0     C     F
1     G  None
2     C     F
3     F     G
4     F  None

In [16]:
# concatenate the dataframes I made with player_stats dataframe
player_stats_with_split_cols = pd.concat([player_stats, split_team_col, split_pos_col], axis=1)

In [17]:
# drop 'team' and 'pos' columns
player_stats_with_split_cols = player_stats_with_split_cols.drop(['team', 'pos'], axis=1)

In [18]:
pd.set_option('display.max_columns', None)

display(player_stats_with_split_cols.head())
display(quintets.head())

player   season  gs  as  age   g    mp    fg   fga    2p  \
0            Bam Adebayo  2023-24  43   1   26  43  34.5   7.6  15.0   7.6   
1          Grayson Allen  2023-24  47   0   28  47  32.9   4.3   8.4   1.7   
2          Jarrett Allen  2023-24  46   0   25  46  30.5   6.4   9.9   6.4   
3  Giannis Antetokounmpo  2023-24  51   1   29  51  35.0  11.5  18.9  11.0   
4             OG Anunoby  2023-24  41   0   26  41  34.1   5.8  11.7   3.7   

    2pa   3p  3pa   ft   fta  orb  drb   trb  ast  stl  blk  tov   pf   pts  \
0  14.8  0.0  0.3  5.0   6.4  2.2  8.3  10.6  4.0  1.0  1.0  2.4  2.5  20.3   
1   3.2  2.6  5.2  1.6   1.8  0.6  3.4   4.0  3.1  0.9  0.6  1.3  2.1  12.8   
2   9.8  0.0  0.1  2.8   3.7  3.6  7.1  10.6  2.8  0.8  1.2  1.8  2.3  15.5   
3  17.1  0.5  1.8  7.3  11.1  2.5  8.7  11.2  6.3  1.4  1.1  3.5  3.1  30.7   
4   6.1  2.1  5.5  1.6   2.0  1.0  3.2   4.2  2.3  1.3  0.7  1.5  2.3  15.3   

    fg%   2p%   3p%   ft%   ts%  efg%   ws  ortg  drtg  ows  dws  ws/48  obpm  \
0  50.5  51.3   9.1  78.0  56.6  50.6  3.9   113   110  1.4  2.5  0.126   0.4   
1  51.1  54.7  49.0  88.4  69.3  66.3  4.1   131   118  2.8  1.3  0.128   0.9   
2  64.4  65.0   0.0  74.7  67.3  64.4  6.6   132   108  3.9  2.8  0.228   1.9   
3  60.9  64.6  25.0  65.7  64.6  62.0  8.6   125   113  6.1  2.5  0.232   6.1   
4  49.8  60.6  37.9  77.4  60.9  58.8  2.5   116   117  1.3  1.3  0.087  -0.6   

   dbpm  bpm  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  \
0   1.4  1.8   1.4  19.5   7.1  27.9  17.3  19.4   1.5   2.8  11.9  26.0   
1  -0.1  0.9   1.1  13.5   2.4  11.1   7.0  12.5   1.3   1.7  12.7  13.9   
2   1.7  3.6   2.0  22.1  13.2  25.0  19.3  13.4   1.3   3.4  13.4  18.9   
3   2.1  8.2   4.6  29.2   8.2  26.4  17.6  30.5   1.9   2.5  12.9  33.1   
4   0.5 -0.1   0.7  13.5   3.1  10.6   6.8   9.4   1.9   1.8  10.7  17.6   

   played_for pos_1 pos_2  
0           1     C     F  
1           1     G  None  
2           1     C     F  
3           1     F     G  
4           2     F  None

season       first_team      second_team      third_team
0  1989-90      Karl Malone       Larry Bird    James Worthy
1  1989-90  Charles Barkley     Tom Chambers    Chris Mullin
2  1989-90    Patrick Ewing  Hakeem Olajuwon  David Robinson
3  1989-90    Magic Johnson    John Stockton   Clyde Drexler
4  1989-90   Michael Jordan    Kevin Johnson      Joe Dumars

My next objective is to resume the whole quintets dataframe into a column. The result would be a column that has value 1, 2 or 3 if the player was in the first, second or third all nba team respectively that season and value 0 if the player wasn't on the all nba teams that season

In [19]:
# I concatenate the text in 'season' column with the other columns
first_team_quintet = quintets['first_team']+quintets['season']
second_team_quintet = quintets['second_team']+quintets['season']
third_team_quintet = quintets['third_team']+quintets['season']

# convert to dataframe
first_team_quintet = first_team_quintet.to_frame()
second_team_quintet = second_team_quintet.to_frame()
third_team_quintet = third_team_quintet.to_frame()

In [20]:
# adding a column labeling the all nba team
first_team_quintet['label'] = 1
second_team_quintet['label'] = 2
third_team_quintet['label'] = 3

# rename first column with the same name in all dataframes
first_team_quintet = first_team_quintet.rename(columns={0:'player_season'})
second_team_quintet = second_team_quintet.rename(columns={0:'player_season'})
third_team_quintet = third_team_quintet.rename(columns={0:'player_season'})

display(first_team_quintet.head())
display(second_team_quintet.head())
display(third_team_quintet.head())

player_season  label
0      Karl Malone1989-90      1
1  Charles Barkley1989-90      1
2    Patrick Ewing1989-90      1
3    Magic Johnson1989-90      1
4   Michael Jordan1989-90      1

player_season  label
0       Larry Bird1989-90      2
1     Tom Chambers1989-90      2
2  Hakeem Olajuwon1989-90      2
3    John Stockton1989-90      2
4    Kevin Johnson1989-90      2

player_season  label
0    James Worthy1989-90      3
1    Chris Mullin1989-90      3
2  David Robinson1989-90      3
3   Clyde Drexler1989-90      3
4      Joe Dumars1989-90      3

In [21]:
# concat the dataframes vertically
labeled_all_nba = pd.concat([first_team_quintet, second_team_quintet, third_team_quintet], ignore_index=True)
labeled_all_nba

player_season  label
0         Karl Malone1989-90      1
1     Charles Barkley1989-90      1
2       Patrick Ewing1989-90      1
3       Magic Johnson1989-90      1
4      Michael Jordan1989-90      1
..                       ...    ...
505      LeBron James2022-23      3
506     Julius Randle2022-23      3
507  Domantas Sabonis2022-23      3
508      De'Aaron Fox2022-23      3
509    Damian Lillard2022-23      3

[510 rows x 2 columns]

In [22]:
# concatenate 'player' and 'season' columns from the player_stats_with_split_cols dataframe into a new dataframe
player_season = player_stats_with_split_cols['player']+player_stats_with_split_cols['season']
player_season = player_season.to_frame()

# rename column
player_season.rename(columns={0: 'player_season'}, inplace=True)

player_season

player_season
0               Bam Adebayo2023-24
1             Grayson Allen2023-24
2             Jarrett Allen2023-24
3     Giannis Antetokounmpo2023-24
4                OG Anunoby2023-24
...                            ...
5159      Dominique Wilkins1989-90
5160         Gerald Wilkins1989-90
5161          Buck Williams1989-90
5162           Kevin Willis1989-90
5163           James Worthy1989-90

[5164 rows x 1 columns]

In [23]:
# left join to label the player_season_corrected dataframe with the all nba team the player has been voted for that season
# in a tryout dataframe
tryout = pd.merge(left = player_season,
                  right = labeled_all_nba,
                  how = 'left', 
                  left_on = "player_season", 
                  right_on= "player_season")
tryout

player_season  label
0               Bam Adebayo2023-24    NaN
1             Grayson Allen2023-24    NaN
2             Jarrett Allen2023-24    NaN
3     Giannis Antetokounmpo2023-24    NaN
4                OG Anunoby2023-24    NaN
...                            ...    ...
5159      Dominique Wilkins1989-90    NaN
5160         Gerald Wilkins1989-90    NaN
5161          Buck Williams1989-90    NaN
5162           Kevin Willis1989-90    NaN
5163           James Worthy1989-90    3.0

[5164 rows x 2 columns]

In [24]:
tryout['label'].value_counts()

# there are combinations of player-season that are not matching

label
2.0    170
1.0    168
3.0    166
Name: count, dtype: int64

In [25]:
# check in which seasons the all nba third team is not complete
pd.set_option('display.max_rows', None)
tryout[tryout['label']==3]

# there is one player missing in 2007-08, 2003-04, 1996-97 and 1994-95.

player_season  label
202         De'Aaron Fox2022-23    3.0
235         LeBron James2022-23    3.0
246       Damian Lillard2022-23    3.0
282        Julius Randle2022-23    3.0
286     Domantas Sabonis2022-23    3.0
397         LeBron James2021-22    3.0
424           Chris Paul2021-22    3.0
440        Pascal Siakam2021-22    3.0
447   Karl-Anthony Towns2021-22    3.0
462           Trae Young2021-22    3.0
480         Bradley Beal2020-21    3.0
493         Jimmy Butler2020-21    3.0
517          Paul George2020-21    3.0
518          Rudy Gobert2020-21    3.0
538         Kyrie Irving2020-21    3.0
634         Jimmy Butler2019-20    3.0
661          Rudy Gobert2019-20    3.0
729          Ben Simmons2019-20    3.0
732         Jayson Tatum2019-20    3.0
744    Russell Westbrook2019-20    3.0
806          Rudy Gobert2018-19    3.0
814        Blake Griffin2018-19    3.0
834         LeBron James2018-19    3.0
891         Kemba Walker2018-19    3.0
893    Russell Westbrook2018-19    3.0
929         Jimmy Butler2017-18    3.0
942        Stephen Curry2017-18    3.0
957          Paul George2017-18    3.0
1018      Victor Oladipo2017-18    3.0
1044  Karl-Anthony Towns2017-18    3.0
1078        Jimmy Butler2016-17    3.0
1095       DeMar DeRozan2016-17    3.0
1112      Draymond Green2016-17    3.0
1138      DeAndre Jordan2016-17    3.0
1197           John Wall2016-17    3.0
1209   LaMarcus Aldridge2015-16    3.0
1246      Andre Drummond2015-16    3.0
1261         Paul George2015-16    3.0
1299          Kyle Lowry2015-16    3.0
1346       Klay Thompson2015-16    3.0
1402          Tim Duncan2014-15    3.0
1422       Blake Griffin2014-15    3.0
1437        Kyrie Irving2014-15    3.0
1445      DeAndre Jordan2014-15    3.0
1496       Klay Thompson2014-15    3.0
1513   LaMarcus Aldridge2013-14    3.0
1549        Goran Dragić2013-14    3.0
1565         Paul George2013-14    3.0
1590        Al Jefferson2013-14    3.0
1606      Damian Lillard2013-14    3.0
1710         Paul George2012-13    3.0
1717        James Harden2012-13    3.0
1729       Dwight Howard2012-13    3.0
1749           David Lee2012-13    3.0
1801         Dwyane Wade2012-13    3.0
1820     Carmelo Anthony2011-12    3.0
1841      Tyson Chandler2011-12    3.0
1914       Dirk Nowitzki2011-12    3.0
1925         Rajon Rondo2011-12    3.0
1939         Dwyane Wade2011-12    3.0
1948   LaMarcus Aldridge2010-11    3.0
2004       Manu Ginóbili2010-11    3.0
2020          Al Horford2010-11    3.0
2068          Chris Paul2010-11    3.0
2071       Zach Randolph2010-11    3.0
2118        Andrew Bogut2009-10    3.0
2146          Tim Duncan2009-10    3.0
2158           Pau Gasol2009-10    3.0
2191         Joe Johnson2009-10    3.0
2233         Brandon Roy2009-10    3.0
2261     Carmelo Anthony2008-09    3.0
2272    Chauncey Billups2008-09    3.0
2309           Pau Gasol2008-09    3.0
2371    Shaquille O'Neal2008-09    3.0
2373         Tony Parker2008-09    3.0
2430       Carlos Boozer2007-08    3.0
2504       Tracy McGrady2007-08    3.0
2509            Yao Ming2007-08    3.0
2529         Paul Pierce2007-08    3.0
2568     Carmelo Anthony2006-07    3.0
2577    Chauncey Billups2006-07    3.0
2616       Kevin Garnett2006-07    3.0
2635       Dwight Howard2006-07    3.0
2709         Dwyane Wade2006-07    3.0
2725     Carmelo Anthony2005-06    3.0
2726      Gilbert Arenas2005-06    3.0
2790       Allen Iverson2005-06    3.0
2814        Shawn Marion2005-06    3.0
2822            Yao Ming2005-06    3.0
2881      Gilbert Arenas2004-05    3.0
2895         Kobe Bryant2004-05    3.0
2966        Shawn Marion2004-05    3.0
2969       Tracy McGrady2004-05    3.0
3022         Ben Wallace2004-05    3.0
3064         Baron Davis2003-04    3.0
3126            Yao Ming2003-04    3.0
3131       Dirk Nowitzki2003-04    3.0
3145        Michael Redd2003-04    3.0
3260     Stephon Marbury2002-03    3.0
3264      Jamal Mashburn2002-03    3.0
3277          Steve Nash2002-03    3.0
3283     Jermaine O'Neal2002-03    3.0
3

In [26]:
# I check in the cleaned dataframe those years
quintets[(quintets['season']=='2007-08')|(quintets['season']=='2003-04')|(quintets['season']=='1996-97')|(quintets['season']=='1994-95')]

# Missing
# 2007-08: Manu Ginóbili (played only 23 games as starter)
# 2003-04: Ron Artest (I found that he changed his name to Metta World Peace in Sep 2011)
# 1996-97: Penny Hardaway (Penny is his nickname. His name is Anfernee Hardaway)
# 1994-95: Dennis Rodman (played only 26 games as starter)

# I won't take into account the 2007-08 season of Manu Ginóbili and the 1994-95 season of Dennis Rodman
# as there are the only 2 players from 510 (15 players each season for the 34 seasons I am analysing)
# that have made the all nba teams without starting more than 35 games.

season        first_team        second_team        third_team
25  1994-95       Karl Malone    Charles Barkley   Detlef Schrempf
26  1994-95    Scottie Pippen         Shawn Kemp     Dennis Rodman
27  1994-95    David Robinson   Shaquille O'Neal   Hakeem Olajuwon
28  1994-95     John Stockton        Gary Payton     Reggie Miller
29  1994-95    Penny Hardaway     Mitch Richmond     Clyde Drexler
35  1996-97       Karl Malone     Scottie Pippen     Anthony Mason
36  1996-97        Grant Hill          Glen Rice         Vin Baker
37  1996-97   Hakeem Olajuwon      Patrick Ewing  Shaquille O'Neal
38  1996-97    Michael Jordan        Gary Payton     John Stockton
39  1996-97      Tim Hardaway     Mitch Richmond    Penny Hardaway
70  2003-04     Kevin Garnett    Jermaine O'Neal     Dirk Nowitzki
71  2003-04        Tim Duncan    Peja Stojaković        Ron Artest
72  2003-04  Shaquille O'Neal        Ben Wallace          Yao Ming
73  2003-04       Kobe Bryant        Sam Cassell      Michael Redd
74  2003-04        Jason Kidd      Tracy McGrady       Baron Davis
90  2007-08     Kevin Garnett      Dirk Nowitzki     Carlos Boozer
91  2007-08      LeBron James         Tim Duncan       Paul Pierce
92  2007-08     Dwight Howard  Amar'e Stoudemire          Yao Ming
93  2007-08       Kobe Bryant         Steve Nash     Tracy McGrady
94  2007-08        Chris Paul     Deron Williams     Manu Ginóbili

In [27]:
# check in which seasons the all nba first team is not complete
pd.set_option('display.max_rows', None)
tryout[tryout['label']==1]

# there is one player missing in 1995-96 and 1994-95.

player_season  label
157     Giannis Antetokounmpo2022-23    1.0
195               Luka Dončić2022-23    1.0
200               Joel Embiid2022-23    1.0
208   Shai Gilgeous-Alexander2022-23    1.0
295              Jayson Tatum2022-23    1.0
317     Giannis Antetokounmpo2021-22    1.0
334              Devin Booker2021-22    1.0
358               Luka Dončić2021-22    1.0
399              Nikola Jokić2021-22    1.0
445              Jayson Tatum2021-22    1.0
469     Giannis Antetokounmpo2020-21    1.0
502             Stephen Curry2020-21    1.0
506               Luka Dončić2020-21    1.0
542              Nikola Jokić2020-21    1.0
548             Kawhi Leonard2020-21    1.0
613     Giannis Antetokounmpo2019-20    1.0
642             Anthony Davis2019-20    1.0
645               Luka Dončić2019-20    1.0
668              James Harden2019-20    1.0
684              LeBron James2019-20    1.0
755     Giannis Antetokounmpo2018-19    1.0
784             Stephen Curry2018-19    1.0
803               Paul George2018-19    1.0
816              James Harden2018-19    1.0
835              Nikola Jokić2018-19    1.0
943             Anthony Davis2017-18    1.0
950              Kevin Durant2017-18    1.0
967              James Harden2017-18    1.0
987              LeBron James2017-18    1.0
999            Damian Lillard2017-18    1.0
1091            Anthony Davis2016-17    1.0
1115             James Harden2016-17    1.0
1135             LeBron James2016-17    1.0
1143            Kawhi Leonard2016-17    1.0
1199        Russell Westbrook2016-17    1.0
1240            Stephen Curry2015-16    1.0
1285             LeBron James2015-16    1.0
1289           DeAndre Jordan2015-16    1.0
1294            Kawhi Leonard2015-16    1.0
1355        Russell Westbrook2015-16    1.0
1395            Stephen Curry2014-15    1.0
1396            Anthony Davis2014-15    1.0
1413               Marc Gasol2014-15    1.0
1423             James Harden2014-15    1.0
1439             LeBron James2014-15    1.0
1554             Kevin Durant2013-14    1.0
1572             James Harden2013-14    1.0
1589             LeBron James2013-14    1.0
1622              Joakim Noah2013-14    1.0
1628               Chris Paul2013-14    1.0
1680              Kobe Bryant2012-13    1.0
1695               Tim Duncan2012-13    1.0
1696             Kevin Durant2012-13    1.0
1734             LeBron James2012-13    1.0
1772               Chris Paul2012-13    1.0
1834              Kobe Bryant2011-12    1.0
1851             Kevin Durant2011-12    1.0
1879            Dwight Howard2011-12    1.0
1886             LeBron James2011-12    1.0
1919               Chris Paul2011-12    1.0
1973              Kobe Bryant2010-11    1.0
1992             Kevin Durant2010-11    1.0
2021            Dwight Howard2010-11    1.0
2027             LeBron James2010-11    1.0
2075             Derrick Rose2010-11    1.0
2126              Kobe Bryant2009-10    1.0
2147             Kevin Durant2009-10    1.0
2179            Dwight Howard2009-10    1.0
2183             LeBron James2009-10    1.0
2246              Dwyane Wade2009-10    1.0
2279              Kobe Bryant2008-09    1.0
2326            Dwight Howard2008-09    1.0
2333             LeBron James2008-09    1.0
2366            Dirk Nowitzki2008-09    1.0
2403              Dwyane Wade2008-09    1.0
2434              Kobe Bryant2007-08    1.0
2460            Kevin Garnett2007-08    1.0
2477            Dwight Howard2007-08    1.0
2484             LeBron James2007-08    1.0
2525               Chris Paul2007-08    1.0
2588              Kobe Bryant2006-07    1.0
2606               Tim Duncan2006-07    1.0
2671               Steve Nash2006-07    1.0
2674            Dirk Nowitzki2006-07    1.0
2702        Amar'e Stoudemire2006-07    1.0
2743              Kobe Bryant2005-06    1.0
2793             LeBron James2005-06    1.0
2825               Steve Nash2005-06    1.0
2828            Dirk Nowitzki2005-06    1.0
2832         Shaquille O'Neal2005-06    1.0
2916               

In [28]:
# I check in the cleaned dataframe those years
quintets[(quintets['season']=='2007-08')|(quintets['season']=='2003-04')|(quintets['season']=='1996-97')|(quintets['season']=='1994-95')]

# Missing
# 1995-96: Penny Hardaway (Penny is his nickname. His name is Anfernee Hardaway)
# 1994-95: Penny Hardaway (Penny is his nickname. His name is Anfernee Hardaway)


season        first_team        second_team        third_team
25  1994-95       Karl Malone    Charles Barkley   Detlef Schrempf
26  1994-95    Scottie Pippen         Shawn Kemp     Dennis Rodman
27  1994-95    David Robinson   Shaquille O'Neal   Hakeem Olajuwon
28  1994-95     John Stockton        Gary Payton     Reggie Miller
29  1994-95    Penny Hardaway     Mitch Richmond     Clyde Drexler
35  1996-97       Karl Malone     Scottie Pippen     Anthony Mason
36  1996-97        Grant Hill          Glen Rice         Vin Baker
37  1996-97   Hakeem Olajuwon      Patrick Ewing  Shaquille O'Neal
38  1996-97    Michael Jordan        Gary Payton     John Stockton
39  1996-97      Tim Hardaway     Mitch Richmond    Penny Hardaway
70  2003-04     Kevin Garnett    Jermaine O'Neal     Dirk Nowitzki
71  2003-04        Tim Duncan    Peja Stojaković        Ron Artest
72  2003-04  Shaquille O'Neal        Ben Wallace          Yao Ming
73  2003-04       Kobe Bryant        Sam Cassell      Michael Redd
74  2003-04        Jason Kidd      Tracy McGrady       Baron Davis
90  2007-08     Kevin Garnett      Dirk Nowitzki     Carlos Boozer
91  2007-08      LeBron James         Tim Duncan       Paul Pierce
92  2007-08     Dwight Howard  Amar'e Stoudemire          Yao Ming
93  2007-08       Kobe Bryant         Steve Nash     Tracy McGrady
94  2007-08        Chris Paul     Deron Williams     Manu Ginóbili

In [29]:
# define a pattern that 'Anfernee' at the beggining
pattern_season = r"^Anfernee"

# use str.contains() to check which values in the column match the pattern
player_stats_with_split_cols[player_stats_with_split_cols['player'].str.contains(pattern_season, na=False)]

player   season  gs  as  age   g    mp   fg   fga   2p   2pa  \
123     Anfernee Simons  2023-24  29   0   24  29  33.6  8.1  18.3  4.7   9.6   
290     Anfernee Simons  2022-23  62   0   23  62  35.0  7.5  16.9  4.1   7.8   
3231  Anfernee Hardaway  2002-03  51   0   31  58  30.6  4.4   9.9  4.0   8.6   
3376  Anfernee Hardaway  2001-02  55   0   30  80  30.8  4.9  11.6  4.6  10.6   
3667  Anfernee Hardaway  1999-00  60   0   28  60  37.6  6.3  13.3  5.8  11.6   
3800  Anfernee Hardaway  1998-99  50   0   27  50  38.9  6.0  14.3  5.2  11.5   
4074  Anfernee Hardaway  1996-97  59   1   25  59  37.6  7.1  15.9  5.7  11.4   
4228  Anfernee Hardaway  1995-96  82   1   24  82  36.8  7.6  14.8  6.5  11.4   
4368  Anfernee Hardaway  1994-95  77   1   23  77  37.7  7.6  14.8  6.5  11.6   
4513  Anfernee Hardaway  1993-94  82   0   22  82  36.8  6.2  13.3  5.6  11.0   

       3p  3pa   ft  fta  orb  drb  trb  ast  stl  blk  tov   pf   pts   fg%  \
123   3.4  8.8  3.8  4.2  0.5  3.0  3.5  5.3  0.5  0.1  2.5  2.0  23.4  44.2   
290   3.4  9.1  2.6  2.9  0.3  2.3  2.6  4.1  0.7  0.2  2.1  2.3  21.1  44.7   
3231  0.4  1.3  1.3  1.7  1.1  3.3  4.4  4.1  1.1  0.4  2.5  2.6  10.6  44.7   
3376  0.3  1.0  2.0  2.4  1.2  3.2  4.4  4.1  1.5  0.4  2.4  2.3  12.0  41.8   
3667  0.6  1.7  3.8  4.8  1.5  4.3  5.8  5.3  1.6  0.6  2.6  2.7  16.9  47.4   
3800  0.8  2.8  3.0  4.2  1.5  4.2  5.7  5.3  2.2  0.5  3.0  2.2  15.8  42.0   
4074  1.4  4.5  4.8  5.8  1.4  3.1  4.5  5.6  1.6  0.6  2.5  2.1  20.5  44.7   
4228  1.1  3.5  5.4  7.1  1.6  2.7  4.3  7.1  2.0  0.5  2.8  2.0  21.7  51.3   
4368  1.1  3.2  4.6  6.0  1.8  2.6  4.4  7.2  1.7  0.3  3.4  2.1  20.9  51.2   
4513  0.6  2.3  3.0  4.0  2.3  3.0  5.4  6.6  2.3  0.6  3.6  2.5  16.0  46.6   

       2p%   3p%   ft%   ts%  efg%    ws  ortg  drtg   ows  dws  ws/48  obpm  \
123   49.1  38.8  90.9  58.0  53.5   1.7   115   122   1.3  0.4  0.084   3.4   
290   52.8  37.7  89.4  58.1  54.8   2.5   113   123   2.1  0.3  0.054   1.3   
3231  46.0  35.6  79.4  49.9  46.9   2.3    97   103   0.4  1.9  0.062  -1.0   
3376  43.2  27.7  81.0  47.2  43.0   2.9    96   104   0.1  2.8  0.056  -1.3   
3667  49.6  32.4  79.0  54.9  49.4   7.8   110   100   4.2  3.6  0.165   1.9   
3800  45.2  28.6  70.6  48.8  44.8   5.0   100    97   1.6  3.4  0.122   0.8   
4074  49.9  31.8  82.0  55.4  49.3   8.1   115   107   5.8  2.2  0.175   4.5   
4228  57.3  31.4  76.7  60.5  54.9  14.4   122   107  11.0  3.4  0.229   6.2   
4368  55.8  34.9  76.9  59.9  55.0  10.7   119   110   8.2  2.5  0.177   4.4   
4513  50.7  26.7  74.2  53.1  48.9   7.1   107   106   3.8  3.3  0.112   1.9   

      dbpm  bpm  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  \
123   -2.3  1.2   0.8  17.3   1.6  10.7   5.9  27.6   0.7   0.4  11.1  28.8   
290   -2.5 -1.1   0.5  14.8   0.9   7.6   4.3  18.6   0.9   0.6  10.2  25.1   
3231   1.0  0.0   0.9  13.0   4.2  12.3   8.2  21.5   1.9   1.1  19.1  19.2   
3376   0.8 -0.4   1.0  13.9   4.4  11.7   8.0  21.1   2.6   0.9  15.7  21.9   
3667   1.6  3.5   3.1  17.9   4.6  12.4   8.6  22.8   2.1   1.2  14.2  21.1   
3800   2.0  2.8   2.3  16.7   4.3  12.8   8.5  25.1   3.1   0.9  15.6  22.4   
4074   0.4  4.9   3.9  21.4   4.2   9.8   6.9  28.5   2.3   1.2  11.7  25.7   
4228   1.0  7.2   6.9  24.6   5.1   8.5   6.8  32.2   2.9   1.0  13.5  25.5   
4368  -0.3  4.1   4.5  20.8   5.7   7.5   6.6  28.5   2.3   0.6  16.1  23.7   
4513   1.0  2.9   3.8  17.4   7.2   9.1   8.2  26.6   3.2   1.1  19.1  21.7   

      played_for pos_1 pos_2  
123            1     G  None  
290            1     G  None  
3231           1     G     F  
3376           1     G     F  
3667           1     G     F  
3800           1     G     F  
4074           1     G     F  
4228           1     G     F  
4368           1     G     F  
4513           1     G     F

In [30]:
# replace on player column of player_stats_with_split_cols dataframe 'Anfernee Hardaway' with 'Penny Hardaway'
# in basketball he was called 'Penny' so I prefer to keep that name.
player_stats_with_split_cols['player'].replace("Anfernee Hardaway", "Penny Hardaway", inplace=True)

In [31]:
# define a pattern that 'Metta' at the beggining
pattern_season = r"^Metta"

# use str.contains() to check which values in the column match the pattern
player_stats_with_split_cols[player_stats_with_split_cols['player'].str.contains(pattern_season, na=False)]

player   season  gs  as  age   g    mp   fg   fga   2p   2pa  \
1670  Metta World Peace  2012-13  66   0   33  75  33.7  4.4  11.0  2.5   5.5   
1823  Metta World Peace  2011-12  45   0   32  64  26.9  2.9   7.3  2.0   4.3   
1953  Metta World Peace  2010-11  82   0   31  82  29.4  3.2   8.0  2.2   5.2   
2107  Metta World Peace  2009-10  77   0   30  77  33.8  4.0   9.6  2.6   5.8   
2262  Metta World Peace  2008-09  55   0   29  69  35.5  6.0  15.0  3.8   9.5   
2419  Metta World Peace  2007-08  54   0   28  57  38.1  7.6  16.9  6.2  13.0   
2570  Metta World Peace  2006-07  65   0   27  70  37.7  6.5  14.9  5.2  11.1   
2727  Metta World Peace  2005-06  56   0   26  56  39.4  6.3  15.5  4.9  11.0   
3038  Metta World Peace  2003-04  71   1   24  73  37.2  6.4  15.2  5.4  11.9   
3183  Metta World Peace  2002-03  67   0   23  69  33.6  5.2  12.3  4.2   9.2   
3335  Metta World Peace  2001-02  50   0   22  55  29.9  4.9  11.6  3.9   8.5   
3482  Metta World Peace  2000-01  74   0   21  76  31.1  4.3  10.7  3.7   8.8   
3632  Metta World Peace  1999-00  63   0   20  72  31.1  4.3  10.5  3.5   7.9   

       3p  3pa   ft  fta  orb  drb  trb  ast  stl  blk  tov   pf   pts   fg%  \
1670  1.9  5.5  1.7  2.3  1.3  3.6  5.0  1.5  1.6  0.6  1.3  2.6  12.4  40.3   
1823  0.9  3.0  1.1  1.8  1.1  2.3  3.4  2.2  1.1  0.4  1.1  2.1   7.7  39.4   
1953  1.0  2.8  1.1  1.7  1.2  2.0  3.3  2.1  1.5  0.4  1.1  2.2   8.5  39.7   
2107  1.4  3.8  1.7  2.4  1.3  3.0  4.3  3.0  1.4  0.3  1.6  2.1  11.0  41.4   
2262  2.2  5.6  2.8  3.8  0.9  4.3  5.2  3.3  1.5  0.3  2.0  2.2  17.1  40.1   
2419  1.5  3.9  3.7  5.2  1.8  4.0  5.8  3.5  2.3  0.7  2.6  2.8  20.5  45.3   
2570  1.3  3.7  4.4  5.9  1.5  5.0  6.5  3.4  2.1  0.6  2.1  2.9  18.8  44.0   
2727  1.4  4.5  3.7  5.4  1.3  3.8  5.1  3.6  2.2  0.7  2.3  2.9  17.6  40.4   
3038  1.0  3.3  4.4  6.0  1.4  3.9  5.3  3.7  2.1  0.7  2.8  2.7  18.3  42.1   
3183  1.0  3.1  4.0  5.4  1.5  3.8  5.2  2.9  2.3  0.7  2.1  3.5  15.5  42.8   
3335  1.0  3.1  2.5  3.7  1.3  3.6  4.9  2.3  2.6  0.7  2.1  3.9  13.2  42.3   
3482  0.6  1.9  2.8  3.7  0.8  3.1  3.9  3.0  2.0  0.6  2.1  3.3  11.9  40.1   
3632  0.8  2.7  2.6  3.9  0.9  3.4  4.3  2.8  1.7  0.5  2.3  2.2  12.0  40.7   

       2p%   3p%   ft%   ts%  efg%   ws  ortg  drtg  ows  dws  ws/48  obpm  \
1670  46.4  34.2  73.4  51.7  48.8  4.6   105   106  1.9  2.6  0.086  -0.5   
1823  46.0  29.6  61.7  47.8  45.4  2.6   102   105  0.9  1.7  0.073  -1.6   
1953  41.9  35.6  67.6  48.5  46.0  4.4   104   105  1.2  3.2  0.088  -1.4   
2107  45.3  35.5  68.8  51.4  48.4  5.3   106   105  1.8  3.5  0.098  -0.6   
2262  40.2  39.9  74.8  51.2  47.5  5.5   104   103  1.6  4.0  0.108   0.7   
2419  47.5  38.0  71.9  53.5  49.7  4.7   106   108  2.4  2.2  0.103   2.1   
2570  46.8  35.8  74.0  53.8  48.5  6.7   108   105  3.6  3.1  0.122   1.9   
2727  44.3  30.8  67.5  49.1  44.9  3.7    99   103  0.5  3.2  0.081  -0.2   
3038  45.2  31.0  73.3  51.1  45.5  8.0   101    96  2.8  5.2  0.141   1.6   
3183  45.8  33.6  73.6  52.9  47.0  7.2   105    98  3.2  4.1  0.150   0.9   
3335  46.4  31.2  66.7  50.1  46.5  2.7    96   100  0.0  2.7  0.079  -0.4   
3482  42.6  29.1  75.0  48.3  42.8  1.9    94   104 -0.1  2.0  0.039  -1.9   
3632  43.8  31.4  67.4  49.1  44.7  2.3    95   103 -0.3  2.6  0.050  -1.1   

      dbpm  bpm  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  \
1670   0.9  0.4   1.5  12.5   4.5  11.5   8.1   6.9   2.5   1.2   9.8  17.5   
1823   1.0 -0.7   0.6  11.0   4.8   9.1   7.1  12.5   2.1   1.1  11.8  15.6   
1953   1.5  0.2   1.3  11.2   4.9   7.6   6.3  10.7   2.7   1.1  11.1  15.3   
2107   1.1  0.5   1.6  12.1   4.3   9.9   7.1  13.3   2.1   0.6  12.8  16.2   
2262   1.2  1.9   2.4  15.6   3.2  13.4   8.5  16.2   2.3   0.7  10.8  24.5   
2419   0.7  2.8   2.6  18.8   5.6  12.1   8.9  16.2   3.1   1.4  11.9  25.5   
2570   0.9  2.8   3.2  18.5   4.4  16.0  10.1  15.6   2.9   1.3  10.7  23.3   
2727   1.3  1.1   1.7  15.2   4.0  1

In [32]:
# replace on player column of player_stats_with_split_cols dataframe 'Metta World Peace' with 'Ron Artest'
# in basketball he was called 'Penny' so I prefer to keep that name.
player_stats_with_split_cols['player'].replace("Metta World Peace", "Ron Artest", inplace=True)

In [33]:
# redo the concatenated dataframe
player_season_corrected = player_stats_with_split_cols['player']+player_stats_with_split_cols['season']
player_season_corrected = player_season_corrected.to_frame()

# change column name
player_season_corrected.rename(columns={0: 'player_season'}, inplace=True)

player_season_corrected.head()

player_season
0            Bam Adebayo2023-24
1          Grayson Allen2023-24
2          Jarrett Allen2023-24
3  Giannis Antetokounmpo2023-24
4             OG Anunoby2023-24

In [34]:
# left join to label the player_season_corrected dataframe with the all nba team the player has been voted for that season
# in a tryout dataframe
tryout_2 = pd.merge(left = player_season_corrected,
                  right = labeled_all_nba,
                  how = 'left', 
                  left_on = "player_season", 
                  right_on= "player_season")
tryout_2

player_season  label
0                  Bam Adebayo2023-24    NaN
1                Grayson Allen2023-24    NaN
2                Jarrett Allen2023-24    NaN
3        Giannis Antetokounmpo2023-24    NaN
4                   OG Anunoby2023-24    NaN
5                  Deni Avdija2023-24    NaN
6                Deandre Ayton2023-24    NaN
7               Paolo Banchero2023-24    NaN
8                 Desmond Bane2023-24    NaN
9              Harrison Barnes2023-24    NaN
10              Scottie Barnes2023-24    NaN
11                  RJ Barrett2023-24    NaN
12               Nicolas Batum2023-24    NaN
13                Bradley Beal2023-24    NaN
14               Malik Beasley2023-24    NaN
15                  Saddiq Bey2023-24    NaN
16                Goga Bitadze2023-24    NaN
17             Bismack Biyombo2023-24    NaN
18               Anthony Black2023-24    NaN
19            Bojan Bogdanović2023-24    NaN
20                Devin Booker2023-24    NaN
21               Mikal Bridges2023-24    NaN
22               Miles Bridges2023-24    NaN
23             Malcolm Brogdon2023-24    NaN
24               Dillon Brooks2023-24    NaN
25                 Bruce Brown2023-24    NaN
26                Jaylen Brown2023-24    NaN
27               Jalen Brunson2023-24    NaN
28                Jimmy Butler2023-24    NaN
29    Kentavious Caldwell-Pope2023-24    NaN
30              Toumani Camara2023-24    NaN
31                Clint Capela2023-24    NaN
32                 Alex Caruso2023-24    NaN
33           Julian Champagnie2023-24    NaN
34                 Nic Claxton2023-24    NaN
35                John Collins2023-24    NaN
36                 Mike Conley2023-24    NaN
37             Cade Cunningham2023-24    NaN
38               Stephen Curry2023-24    NaN
39               Anthony Davis2023-24    NaN
40               DeMar DeRozan2023-24    NaN
41           Spencer Dinwiddie2023-24    NaN
42            Donte DiVincenzo2023-24    NaN
43                 Luka Dončić2023-24    NaN
44               Luguentz Dort2023-24    NaN
45                   Kris Dunn2023-24    NaN
46                Kevin Durant2023-24    NaN
47                 Jalen Duren2023-24    NaN
48             Anthony Edwards2023-24    NaN
49                 Joel Embiid2023-24    NaN
50         Dorian Finney-Smith2023-24    NaN
51           Simone Fontecchio2023-24    NaN
52                De'Aaron Fox2023-24    NaN
53              Daniel Gafford2023-24    NaN
54              Darius Garland2023-24    NaN
55                 Paul George2023-24    NaN
56                 Josh Giddey2023-24    NaN
57     Shai Gilgeous-Alexander2023-24    NaN
58                 Rudy Gobert2023-24    NaN
59                Aaron Gordon2023-24    NaN
60                Jerami Grant2023-24    NaN
61                 Jalen Green2023-24    NaN
62           Tyrese Haliburton2023-24    NaN
63                James Harden2023-24    NaN
64               Tobias Harris2023-24    NaN
65               Killian Hayes2023-24    NaN
66              Gordon Hayward2023-24    NaN
67                 Tyler Herro2023-24    NaN
68                 Buddy Hield2023-24    NaN
69                Jrue Holiday2023-24    NaN
70               Chet Holmgren2023-24    NaN
71               Kevin Huerter2023-24    NaN
72              Brandon Ingram2023-24    NaN
73                Kyrie Irving2023-24    NaN
74                  Jaden Ivey2023-24    NaN
75           Jaren Jackson Jr.2023-24    NaN
76                LeBron James2023-24    NaN
77             Cameron Johnson2023-24    NaN
78               Jalen Johnson2023-24    NaN
79              Keldon Johnson2023-24    NaN
80                Nikola Jokić2023-24    NaN
81           Derrick Jones Jr.2023-24    NaN
82               Herbert Jones2023-24    NaN
83                  Tyus Jones2023-24    NaN
84                  Kyle Kuzma2023-24    NaN
85               Kawhi Leonard2023-24    NaN
86              Damian Lillard2023-24    NaN
87            Dereck Lively II2023-24    NaN
88                K

In [35]:
tryout_2['label'].value_counts()

# now it is all right

label
1.0    170
2.0    170
3.0    168
Name: count, dtype: int64

In [36]:
# check the shape of the dataframes

display(tryout_2.shape)
player_stats_with_split_cols.shape

(5164, 2)

(5164, 52)

In [37]:
# check the index

display(tryout_2.head())
display(player_stats_with_split_cols.head())

display(tryout_2.tail())
player_stats_with_split_cols.tail()

player_season  label
0            Bam Adebayo2023-24    NaN
1          Grayson Allen2023-24    NaN
2          Jarrett Allen2023-24    NaN
3  Giannis Antetokounmpo2023-24    NaN
4             OG Anunoby2023-24    NaN

player   season  gs  as  age   g    mp    fg   fga    2p  \
0            Bam Adebayo  2023-24  43   1   26  43  34.5   7.6  15.0   7.6   
1          Grayson Allen  2023-24  47   0   28  47  32.9   4.3   8.4   1.7   
2          Jarrett Allen  2023-24  46   0   25  46  30.5   6.4   9.9   6.4   
3  Giannis Antetokounmpo  2023-24  51   1   29  51  35.0  11.5  18.9  11.0   
4             OG Anunoby  2023-24  41   0   26  41  34.1   5.8  11.7   3.7   

    2pa   3p  3pa   ft   fta  orb  drb   trb  ast  stl  blk  tov   pf   pts  \
0  14.8  0.0  0.3  5.0   6.4  2.2  8.3  10.6  4.0  1.0  1.0  2.4  2.5  20.3   
1   3.2  2.6  5.2  1.6   1.8  0.6  3.4   4.0  3.1  0.9  0.6  1.3  2.1  12.8   
2   9.8  0.0  0.1  2.8   3.7  3.6  7.1  10.6  2.8  0.8  1.2  1.8  2.3  15.5   
3  17.1  0.5  1.8  7.3  11.1  2.5  8.7  11.2  6.3  1.4  1.1  3.5  3.1  30.7   
4   6.1  2.1  5.5  1.6   2.0  1.0  3.2   4.2  2.3  1.3  0.7  1.5  2.3  15.3   

    fg%   2p%   3p%   ft%   ts%  efg%   ws  ortg  drtg  ows  dws  ws/48  obpm  \
0  50.5  51.3   9.1  78.0  56.6  50.6  3.9   113   110  1.4  2.5  0.126   0.4   
1  51.1  54.7  49.0  88.4  69.3  66.3  4.1   131   118  2.8  1.3  0.128   0.9   
2  64.4  65.0   0.0  74.7  67.3  64.4  6.6   132   108  3.9  2.8  0.228   1.9   
3  60.9  64.6  25.0  65.7  64.6  62.0  8.6   125   113  6.1  2.5  0.232   6.1   
4  49.8  60.6  37.9  77.4  60.9  58.8  2.5   116   117  1.3  1.3  0.087  -0.6   

   dbpm  bpm  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  \
0   1.4  1.8   1.4  19.5   7.1  27.9  17.3  19.4   1.5   2.8  11.9  26.0   
1  -0.1  0.9   1.1  13.5   2.4  11.1   7.0  12.5   1.3   1.7  12.7  13.9   
2   1.7  3.6   2.0  22.1  13.2  25.0  19.3  13.4   1.3   3.4  13.4  18.9   
3   2.1  8.2   4.6  29.2   8.2  26.4  17.6  30.5   1.9   2.5  12.9  33.1   
4   0.5 -0.1   0.7  13.5   3.1  10.6   6.8   9.4   1.9   1.8  10.7  17.6   

   played_for pos_1 pos_2  
0           1     C     F  
1           1     G  None  
2           1     C     F  
3           1     F     G  
4           2     F  None

player_season  label
5159  Dominique Wilkins1989-90    NaN
5160     Gerald Wilkins1989-90    NaN
5161      Buck Williams1989-90    NaN
5162       Kevin Willis1989-90    NaN
5163       James Worthy1989-90    3.0

player   season  gs  as  age   g    mp    fg   fga   2p  \
5159  Dominique Wilkins  1989-90  79   1   30  80  36.1  10.1  20.9  9.4   
5160     Gerald Wilkins  1989-90  80   0   26  82  31.8   5.8  12.6  5.3   
5161      Buck Williams  1989-90  82   0   29  82  34.2   5.0   9.2  5.0   
5162       Kevin Willis  1989-90  51   0   27  81  28.1   5.2   9.9  5.1   
5163       James Worthy  1989-90  80   1   28  80  37.0   8.9  16.2  8.7   

       2pa   3p  3pa   ft  fta  orb  drb  trb  ast  stl  blk  tov   pf   pts  \
5159  18.6  0.7  2.3  5.7  7.1  2.7  3.8  6.5  2.5  1.6  0.6  2.2  1.8  26.7   
5160  11.1  0.5  1.5  2.5  3.2  1.6  2.9  4.5  4.0  1.2  0.3  2.4  2.3  14.5   
5161   9.2  0.0  0.0  3.5  5.0  3.0  6.7  9.8  1.4  0.8  0.5  2.0  3.5  13.6   
5162   9.9  0.0  0.1  2.1  3.0  3.1  4.8  8.0  0.7  0.8  0.6  1.8  3.2  12.4   
5163  15.6  0.2  0.6  3.1  4.0  2.0  4.0  6.0  3.6  1.2  0.6  2.0  2.4  21.1   

       fg%   2p%   3p%   ft%   ts%  efg%    ws  ortg  drtg  ows  dws  ws/48  \
5159  48.4  50.4  32.2  80.7  55.6  50.2  11.1   118   112  9.3  1.8  0.184   
5160  45.7  47.7  31.2  80.3  52.0  47.6   4.4   106   110  2.3  2.1  0.082   
5161  54.8  54.8   0.0  70.6  59.7  54.8   8.7   116   104  4.3  4.3  0.149   
5162  51.9  52.1  28.6  68.3  55.1  52.0   4.7   111   111  3.0  1.7  0.098   
5163  54.8  55.7  30.6  78.2  58.6  55.4  10.6   119   108  7.5  3.1  0.172   

      obpm  dbpm  bpm  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  \
5159   6.5  -1.1  5.4   5.4  24.6   8.4  12.4  10.4  11.8   2.2   1.0   8.3   
5160   0.3  -0.6 -0.3   1.1  14.3   5.8   9.8   7.9  18.5   1.8   0.5  14.5   
5161   0.1   0.6  0.7   1.9  14.3   9.6  21.6  15.5   5.5   1.2   0.8  15.3   
5162   0.0  -1.8 -1.7   0.1  15.7  12.5  20.4  16.4   3.7   1.4   1.3  13.6   
5163   3.6   0.2  3.8   4.3  19.8   6.4  11.9   9.3  15.6   1.7   1.0  10.0   

      usg%  played_for pos_1 pos_2  
5159  30.4           1     F     G  
5160  21.4           1     G     F  
5161  15.5           1     F     C  
5162  19.5           1     F     C  
5163  23.0           1     F  None

In [38]:
# add the label column to the player_stats_with_split_cols dataframe
player_stats_with_split_cols['all_nba_team_label'] = tryout_2['label']
player_stats_with_split_cols.head()

# now I just need this table as I have added all the information from the all

player   season  gs  as  age   g    mp    fg   fga    2p  \
0            Bam Adebayo  2023-24  43   1   26  43  34.5   7.6  15.0   7.6   
1          Grayson Allen  2023-24  47   0   28  47  32.9   4.3   8.4   1.7   
2          Jarrett Allen  2023-24  46   0   25  46  30.5   6.4   9.9   6.4   
3  Giannis Antetokounmpo  2023-24  51   1   29  51  35.0  11.5  18.9  11.0   
4             OG Anunoby  2023-24  41   0   26  41  34.1   5.8  11.7   3.7   

    2pa   3p  3pa   ft   fta  orb  drb   trb  ast  stl  blk  tov   pf   pts  \
0  14.8  0.0  0.3  5.0   6.4  2.2  8.3  10.6  4.0  1.0  1.0  2.4  2.5  20.3   
1   3.2  2.6  5.2  1.6   1.8  0.6  3.4   4.0  3.1  0.9  0.6  1.3  2.1  12.8   
2   9.8  0.0  0.1  2.8   3.7  3.6  7.1  10.6  2.8  0.8  1.2  1.8  2.3  15.5   
3  17.1  0.5  1.8  7.3  11.1  2.5  8.7  11.2  6.3  1.4  1.1  3.5  3.1  30.7   
4   6.1  2.1  5.5  1.6   2.0  1.0  3.2   4.2  2.3  1.3  0.7  1.5  2.3  15.3   

    fg%   2p%   3p%   ft%   ts%  efg%   ws  ortg  drtg  ows  dws  ws/48  obpm  \
0  50.5  51.3   9.1  78.0  56.6  50.6  3.9   113   110  1.4  2.5  0.126   0.4   
1  51.1  54.7  49.0  88.4  69.3  66.3  4.1   131   118  2.8  1.3  0.128   0.9   
2  64.4  65.0   0.0  74.7  67.3  64.4  6.6   132   108  3.9  2.8  0.228   1.9   
3  60.9  64.6  25.0  65.7  64.6  62.0  8.6   125   113  6.1  2.5  0.232   6.1   
4  49.8  60.6  37.9  77.4  60.9  58.8  2.5   116   117  1.3  1.3  0.087  -0.6   

   dbpm  bpm  vorp   per  orb%  drb%  trb%  ast%  stl%  blk%  tov%  usg%  \
0   1.4  1.8   1.4  19.5   7.1  27.9  17.3  19.4   1.5   2.8  11.9  26.0   
1  -0.1  0.9   1.1  13.5   2.4  11.1   7.0  12.5   1.3   1.7  12.7  13.9   
2   1.7  3.6   2.0  22.1  13.2  25.0  19.3  13.4   1.3   3.4  13.4  18.9   
3   2.1  8.2   4.6  29.2   8.2  26.4  17.6  30.5   1.9   2.5  12.9  33.1   
4   0.5 -0.1   0.7  13.5   3.1  10.6   6.8   9.4   1.9   1.8  10.7  17.6   

   played_for pos_1 pos_2  all_nba_team_label  
0           1     C     F                 NaN  
1           1     G  None                 NaN  
2           1     C     F                 NaN  
3           1     F     G                 NaN  
4           2     F  None                 NaN

In [39]:
player_stats_with_split_cols['all_nba_team_label'].value_counts()

all_nba_team_label
1.0    170
2.0    170
3.0    168
Name: count, dtype: int64

In [40]:
player_stats_with_split_cols.isnull().sum()

player                   0
season                   0
gs                       0
as                       0
age                      0
g                        0
mp                       0
fg                       0
fga                      0
2p                       0
2pa                      0
3p                       0
3pa                      0
ft                       0
fta                      0
orb                      0
drb                      0
trb                      0
ast                      0
stl                      0
blk                      0
tov                      0
pf                       0
pts                      0
fg%                      0
2p%                      0
3p%                      0
ft%                      0
ts%                      0
efg%                     0
ws                       0
ortg                     0
drtg                     0
ows                      0
dws                      0
ws/48                    0
obpm                     0
d

In [41]:
# fill the null values with 'NA'
fill_na_list = ['pos_2', 'all_nba_team_label']

for col in fill_na_list:
    player_stats_with_split_cols[col] = player_stats_with_split_cols[col].fillna('NA')

# check again for NaN
player_stats_with_split_cols.isnull().sum().sum()

0

In [42]:
# change the value labels of 'all_nba_team_label'
player_stats_with_split_cols['all_nba_team_label'].replace({1: 'first',
                                                           2: 'second',
                                                           3: 'third'}, inplace=True)

player_stats_with_split_cols['all_nba_team_label'].value_counts()

all_nba_team_label
NA        4656
first      170
second     170
third      168
Name: count, dtype: int64

In [43]:
# save the dataframe player_stats_with_split_cols to csv.
player_stats_with_split_cols.to_csv('player_stats_complete.csv', sep=',')

In [44]:
# push the dataframe to SQL 

player_stats_with_split_cols.to_sql('player_stats_complete', con=engine)

5164